SwiftUIBasics/Views/SignUpView.swift

```diff
@@ -6,40 +6,191 @@
+                    return true
+                } else {
+                    return false
+                }
+            }
+            .assign(to: \.isValidUsernameLength, on: self)
+            .store(in: &cancelableSet)
+
+        $password
+            .receive(on: RunLoop.main)
+            .map { password in
+                return password.count >= 8
+            }
+            .assign(to: \.isValidPasswordLength, on: self)
+            .store(in: &cancelableSet)
+
+        $password
+            .receive(on: RunLoop.main)
+            .map { password in
+                let pattern = "[A-Z]"
+                if let _ = password.range(of: pattern, options: .regularExpression) {
+                    return true
+                } else {
+                    return false
+                }
+            }
+            .assign(to: \.isValidPasswordUpperCase, on: self)
+            .store(in: &cancelableSet)
+        
+        $password
+            .receive(on: RunLoop.main)
+            .map { password in
+                let pattern = "[a-z]"
+                if let _ = password.range(of: pattern, options: .regularExpression) {
+                    return true
+                } else {
+                    return false
+                }
+            }
+            .assign(to: \.isValidPasswordLowerCase, on: self)
+            .store(in: &cancelableSet)
+        
+        $password
+            .receive(on: RunLoop.main)
+            .map { password in
+                let pattern = "[!\"#$%&'()*+,-./:;<=>?@\\[\\\\\\]^_`{|}~]+"
+                if let _ = password.range(of: pattern, options: .regularExpression) {
+                    return true
+                } else {
+                    return false
+                }
+            }
+            .assign(to: \.isValidPassWordOneSymbol, on: self)
+            .store(in: &cancelableSet)
+        
+        $password
+            .receive(on: RunLoop.main)
+            .map { password in
+                let pattern = "[0-9]"
+                if let _ = password.range(of: pattern, options: .regularExpression) {
+                    return true
+                } else {
+                    return false
+                }
+            }
+            .assign(to: \.isValidPasswordOneNumber, on: self)
+            .store(in: &cancelableSet)
+        
+        
+
+        Publishers.CombineLatest($password, $passwordConfirm)
+            .receive(on: RunLoop.main)
+            .map { (password, passwordConfirm) in
+                return !password.isEmpty && !passwordConfirm.isEmpty && password == passwordConfirm
+            }
+            .assign(to: \.isValidPasswordMatch, on: self)
+            .store(in: &cancelableSet)
+
+        Publishers.CombineLatest4($isValidPasswordLowerCase, $isValidPassWordOneSymbol, $isValidPasswordOneNumber, $isValidPre)
+            .map { (a, b, c, d) in
+                return a && b && c && d
+            }
+            .assign(to: \.isValidPre, on: self)
```

Comment:
circular reference you are evaluating isValidPre and assigning it to the same variable